lib/galaxy/tools/cache.py

```diff
@@ -138,59 +132,30 @@ class ToolShedRepositoryCache(object):
 
     def __init__(self, app):
         self.app = app
-        self.cache = local()
+        self.local_repositories = []
+        self.repositories = []
+        self.repos_by_tuple = defaultdict(list)
 
     def add_local_repository(self, repository):
-        self.cache.repositories.append(repository)
-        self.cache.repos_by_tuple[(repository.tool_shed, repository.owner, repository.name)].append(repository)
+        self.local_repositories.append(repository)
+        self.repos_by_tuple[(repository.tool_shed, repository.owner, repository.name)].append(repository)
 
     @property
     def tool_shed_repositories(self):
-        try:
-            repositories = self.cache.repositories
-        except AttributeError:
-            self.rebuild()
-            repositories = self.cache.repositories
-        tool_shed_repositories = [repo for repo in repositories if isinstance(repo, self.app.install_model.ToolShedRepository)]
-        if tool_shed_repositories and inspect(tool_shed_repositories[0]).detached:
-            self.rebuild()
-            repositories = self.cache.repositories
-        return repositories
+        return self.repositories
 
     @property
     def tool_shed_repos_by_tuple(self):
```

Comment:
Since this property is only used once, I'd just drop it and replace it with `self.repos_by_tuple` inside `get_installed_repository()` below.

Reply:
Thanks! Added that in the rebase